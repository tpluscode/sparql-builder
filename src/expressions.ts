export { IN } from './lib/expressions/in'
export { VALUES } from './lib/expressions/values'
