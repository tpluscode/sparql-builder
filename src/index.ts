export { sparql } from '@tpluscode/rdf-string'
export { DELETE } from './lib/DeleteBuilder.js'
export type { DeleteInsertQuery as DeleteInsert, DeleteData } from './lib/DeleteBuilder.js'
export { INSERT } from './lib/InsertBuilder.js'
export type { InsertQuery as Insert, InsertData } from './lib/InsertBuilder.js'
export { ASK } from './lib/AskBuilder.js'
export type { AskQuery as Ask } from './lib/AskBuilder.js'
export { DESCRIBE } from './lib/DescribeBuilder.js'
export type { DescribeQuery as Describe } from './lib/DescribeBuilder.js'
export { CONSTRUCT } from './lib/ConstructBuilder.js'
export type { ConstructQuery as Construct } from './lib/ConstructBuilder.js'
export { SELECT } from './lib/SelectBuilder.js'
export type { SelectQuery as Select } from './lib/SelectBuilder.js'
export { WITH } from './lib/WithBuilder.js'
export { prefixes } from '@tpluscode/rdf-string'
export type { SparqlTemplateResult } from '@tpluscode/rdf-string'
export type {
  SparqlAskExecutable, SparqlGraphQueryExecutable, SparqlQueryExecutable, SparqlUpdateExecutable,
} from './lib/index.js'
