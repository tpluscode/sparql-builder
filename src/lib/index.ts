import { AskQuery, ConstructQuery, QueryOptions, SelectQuery, UpdateQuery } from 'sparql-http-client'
import { SparqlTemplateResult } from '@tpluscode/rdf-string'
import prologue, { PrologueBuilder } from './partials/prologue'

interface SparqlBuildOptions {
  base?: string
}

export type SparqlExecuteOptions = QueryOptions & SparqlBuildOptions

export interface SparqlQuery extends PrologueBuilder {
  build(options?: SparqlBuildOptions): string
  _getTemplateResult(): SparqlTemplateResult
}

export interface SparqlQueryExecutable {
  execute<TQuery extends SelectQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['select']>
}

export interface SparqlGraphQueryExecutable {
  execute<TQuery extends ConstructQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['construct']>
}

export interface SparqlUpdateExecutable {
  execute<TQuery extends UpdateQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['update']>
}

export interface SparqlAskExecutable {
  execute<TQuery extends AskQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['ask']>
}

type TBuilder = Pick<SparqlQuery, 'build'> & Pick<SparqlTemplateResult, '_toPartialString'>

// eslint-disable-next-line no-unused-vars
export default function Builder<T extends SparqlQuery>(): TBuilder & T {
  return {
    ...prologue(),
    build(this: SparqlQuery, { base }: SparqlBuildOptions = {}): string {
      const queryResult = this._getTemplateResult().toString({
        base,
      })

      if (this.prologueResult) {
        return `${this.prologueResult}\n\n${queryResult}`
      }

      return queryResult
    },
    _toPartialString(this: SparqlQuery, options: any) {
      return this._getTemplateResult()._toPartialString(options)
    },
  } as any
}
