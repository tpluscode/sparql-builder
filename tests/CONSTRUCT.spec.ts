import namespace from '@rdfjs/namespace'
import { dbo, foaf, schema, skos } from '@tpluscode/rdf-ns-builders'
import * as RDF from '@rdfjs/data-model'
import { namedNode } from '@rdfjs/data-model'
import { CONSTRUCT, SELECT } from '../src'
import { sparqlClient } from './_mocks'

describe('CONSTRUCT', () => {
  it('executes as construct', () => {
    // given
    const client = sparqlClient()

    // when
    CONSTRUCT``.execute(client)

    // then
    expect(client.construct).toHaveBeenCalled()
  })

  it('can have additional prologue', () => {
    // given
    const base = namedNode('http://foo.bar/baz')

    // when
    const query = CONSTRUCT`?s ?p ?o .`
      .prologue`#pragma join.hash off`
      .prologue`BASE ${base}`
      .build()

    // then
    expect(query).toMatchSnapshot()
  })

  it('generates empty WHERE clause by default', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/>
    CONSTRUCT { <http://example.com/me> a schema:Person } WHERE {}`

    // when
    const actual = CONSTRUCT`<http://example.com/me> a ${schema.Person}`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('support shorthand syntax', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/>
    CONSTRUCT WHERE { ?person a schema:Person }`

    // when
    const actual = CONSTRUCT.WHERE`?person a ${schema.Person}`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('correctly interpolate shorthand syntax', () => {
    // given
    const root = RDF.namedNode('https://example.com/')
    const expected = `PREFIX skos: <${skos().value}>
    CONSTRUCT WHERE { 
      <https://example.com/> skos:narrower ?narrower .
      ?narrower skos:prefLabel ?label . 
    }`

    // when
    const actual = CONSTRUCT.WHERE`
      ${root} ${skos.narrower} ?narrower .
      ?narrower ${skos.prefLabel} ?label .
    `.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('support FROM in shorthand syntax', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/>
    CONSTRUCT FROM <http://example.com/graph> WHERE { ?person a schema:Person }`

    // when
    const actual = CONSTRUCT.WHERE`?person a ${schema.Person}`
      .FROM(RDF.namedNode('http://example.com/graph')).build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports LIMIT/OFFSET', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/>
    CONSTRUCT { <http://example.com/me> a schema:Person } WHERE {}
    LIMIT 5 OFFSET 305`

    // when
    const actual = CONSTRUCT`<http://example.com/me> a ${schema.Person}`.LIMIT(5).OFFSET(305).build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be constructed with a base', () => {
    // given
    const ns = namespace('http://example.com/')
    const expected = `BASE <http://example.com/>

    CONSTRUCT {
      <person> a <Person>
    } WHERE {}`

    // when
    const actual = CONSTRUCT`${ns.person} a ${ns.Person}`.build({
      base: 'http://example.com/',
    })

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be combined with another query to create a subquery', () => {
    // given
    const person = RDF.variable('person')
    const selectPeopleBornInBerlin = SELECT`${person}`
      .WHERE`${person} ${dbo.birthPlace} <http://dbpedia.org/resource/Berlin>`
      .LIMIT(100)
    const expected = `PREFIX dbo: <http://dbpedia.org/ontology/>
PREFIX foaf: <http://xmlns.com/foaf/0.1/>

CONSTRUCT { ?person ?p ?o }
WHERE {

    VALUES ?p { dbo:birthDate dbo:deathDate foaf:name }
    ?person ?p ?o .

    {
        SELECT ?person
        
        WHERE {
            ?person dbo:birthPlace <http://dbpedia.org/resource/Berlin>
        }
        LIMIT 100 
    }
}`

    // when
    const construct = CONSTRUCT`${person} ?p ?o`
      .WHERE`
        VALUES ?p { ${dbo.birthDate} ${dbo.deathDate} ${foaf.name} }
        ${person} ?p ?o .

        ${selectPeopleBornInBerlin}
    `.build()

    // then
    expect(construct).toMatchQuery(expected)
  })

  it('adds FROM when default graph set', () => {
    // given
    const expected = 'CONSTRUCT { ?s ?p ?o } FROM <urn:foo:bar> WHERE { ?s ?p ?o }'

    // when
    const actual = CONSTRUCT`?s ?p ?o`.FROM(RDF.namedNode('urn:foo:bar')).WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports multiple FROM NAMED', () => {
    // given
    const expected = `CONSTRUCT { ?s ?p ?o }
      FROM NAMED <urn:foo:bar>
      FROM NAMED <urn:foo:baz>
      WHERE { ?s ?p ?o }`

    // when
    const actual = CONSTRUCT`?s ?p ?o`
      .FROM().NAMED(RDF.namedNode('urn:foo:bar'))
      .FROM().NAMED(RDF.namedNode('urn:foo:baz'))
      .WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports multiple FROM', () => {
    // given
    const expected = `CONSTRUCT { ?s ?p ?o }
      FROM <urn:foo:bar>
      FROM <urn:foo:baz>
      WHERE { ?s ?p ?o }`

    // when
    const actual = CONSTRUCT`?s ?p ?o`
      .FROM(RDF.namedNode('urn:foo:bar'))
      .FROM(RDF.namedNode('urn:foo:baz'))
      .WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('allows mixing FROM and FROM NAMED', () => {
    // given
    const expected = `CONSTRUCT { ?s ?p ?o }
      FROM <urn:foo:bar>
      FROM NAMED <urn:foo:baz>
      WHERE { ?s ?p ?o }`

    // when
    const actual = CONSTRUCT`?s ?p ?o`
      .FROM(RDF.namedNode('urn:foo:bar'))
      .FROM().NAMED(RDF.namedNode('urn:foo:baz'))
      .WHERE`?s ?p ?o`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })
})
