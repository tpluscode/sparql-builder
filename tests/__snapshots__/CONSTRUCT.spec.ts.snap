// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONSTRUCT can have additional prologue 1`] = `
"#pragma join.hash off
BASE <http://foo.bar/baz>

CONSTRUCT { ?s ?p ?o . }

WHERE {}
"
`;
